vhdbuilder/packer/backfill-cleanup.sh

```diff
@@ -49,28 +49,29 @@ fi
   gallery_list=$(az sig list -g ${AZURE_RESOURCE_GROUP_NAME} | jq -r '.[] | select(.name != "AKSWindows") | .name')
   for gallery in $gallery_list; do
-    # delete old Windows SIG image versions in gallery (image definitions must have .osType == "Windows")
-    image_defs=$(az sig image-definition list -g ${AZURE_RESOURCE_GROUP_NAME} -r ${gallery} | jq -r '.[] | select(.osType == "Windows").name')
-    for image_definition in $image_defs; do
-        echo "Finding sig image versions associated with ${image_definition} in gallery ${gallery}"
-        image_versions=$(az sig image-version list -g ${AZURE_RESOURCE_GROUP_NAME} -r ${gallery} -i ${image_definition} | jq --arg dl $deadline -r '.[] | select(.tags.now < $dl).name')
-        for image_version in $image_versions; do
-            az sig image-version show -e $image_version -i ${image_definition} -r ${gallery} -g ${AZURE_RESOURCE_GROUP_NAME} | jq .id
-            echo "Deleting sig image-version ${image_version} ${image_definition} from gallery ${gallery} rg ${AZURE_RESOURCE_GROUP_NAME}"
-            az sig image-version delete -e $image_version -i ${image_definition} -r ${gallery} -g ${AZURE_RESOURCE_GROUP_NAME}
-        done
-        image_versions=$(az sig image-version list -g ${AZURE_RESOURCE_GROUP_NAME} -r ${gallery} -i ${image_definition})
-        # clean the image-definition if ALL associated sig versions have been deleted
-        if [[ "$image_versions" == "[]" ]]; then
-          echo "Deleting sig image-definition ${image_definition} from gallery ${gallery} rg ${AZURE_RESOURCE_GROUP_NAME}"
-          az sig image-definition delete --gallery-image-definition ${image_definition} -r ${gallery} -g ${AZURE_RESOURCE_GROUP_NAME}
-        fi
-    done
-    image_defs=$(az sig image-definition list -g ${AZURE_RESOURCE_GROUP_NAME} -r ${gallery} | jq -r '.[] | select(.osType == "Windows").name')
-    # clean the gallery if ALL sig image-definitions have been deleted
-    if [[ -z $image_defs ]]; then
+    if [[ "${gallery}" =~ WS2019Gallery* ]]; then
+      create_date=${gallery:13:6}
+    elif [[ "${gallery}" =~ WS2019_containerdGallery* ]]; then
+      create_date=${gallery:24:6}
+    elif [[ "${gallery}" =~ WS2022_containerdGallery* ]]; then
+      create_date=${gallery:24:6}
+    elif [[ "${gallery}" =~ WS2022_containerd_gen2Gallery* ]]; then
+      create_date=${gallery:29:6}
+    else
+      continue
+    fi
+
+    due_date=$(date +%y%m%d -d "7 days ago")
+    echo "create_date is ${create_date}"
+    echo "due_date is ${due_date}"
+    # clean the entire SIG resources if it's one week ago
+    if [[ $create_date < $due_date ]]; then
       echo "Deleting gallery ${gallery}"
       az sig delete --gallery-name ${gallery} --resource-group ${AZURE_RESOURCE_GROUP_NAME}
+      image_defs=$(az sig image-definition list -g ${AZURE_RESOURCE_GROUP_NAME} -r ${gallery} | jq -r '.[] | select(.osType == "Windows").name')
+      if [[ -n "${image_defs}" ]]; then 
```

Comment:
can a gallery be deleted before deleting all nested resources? Should we keep the old logic to delete all image definitions and versions?

Reply:
I think I need to do some manual experiments: create gallery => SIG image => SIG version and delete gallery completely before deleting all nested resources versus bottom-up deletion (_old logic_)

Reply:
short answer is we can't by running this single command alone. error message is "can not delete resource before nested resources are deleted".

![image](https://user-images.githubusercontent.com/103480804/221899775-1751280f-46a6-45e2-9ba0-222a4f782c1d.png)

we can do this on portal which I often did before by typing "yes" when asked if you wanna clean all nested resources, as shown below.
![image](https://user-images.githubusercontent.com/103480804/221900250-3d3fab93-a78c-486b-9a0a-1a95acd93bd5.png)